test/unit_test/cats_tree_tests.cc

```diff
@@ -0,0 +1,516 @@
+  ec.ft_offset = 0;
+  ec._current_reduction_depth = 0;
+  ec.l.cb = CB::label();
+  ec.l.cb.costs.push_back({3.5f, 2, 0.5f, 0.0f});
+  ec.l.cb.costs.push_back({3.5f, 5, 0.5f, 0.0f});
+
+  predictions_t preds_to_return = {-1.f, 1.f, -1.f};
+
+  reduction_test_harness* pharness = nullptr;
+  auto& base = *as_singleline(get_test_harness_reduction(preds_to_return, pharness));
+  cats_tree tree;
+  tree.init(8, 1);
+
+  tree.learn(base, ec);
+
+  // verify 1) # of calls to learn 2) passed in labels 3) passed in weights
+  vector<label_data> expected_labels = {
+    {1, 3.5f / .5f, 0},
+    {1, 3.5f / .5f, 0},
+    {1, 3.5f / .5f, 0},
+   };
+  BOOST_CHECK_EQUAL_COLLECTIONS(
+      pharness->_labels.begin(), pharness->_labels.end(), expected_labels.begin(), expected_labels.end());
+
+  // verify id of learners that were trained
+  vector<uint64_t> expected_learners = {5, 2, 0};
+  BOOST_CHECK_EQUAL_COLLECTIONS(pharness->_learner_offset.begin(), pharness->_learner_offset.end(),
+      expected_learners.begin(), expected_learners.end());
+
+  destroy_free<test_learner_t>(&base);
+  ec.l.cb.costs.delete_v();
+}
+
+BOOST_AUTO_TEST_CASE(offset_tree_cont_predict)
+{
+  // 0 node tree
+  VW::cats_tree::predict_test_helper({}, 0, 0, 0);
+  // 2 node trees
+  VW::cats_tree::predict_test_helper({-1}, 1, 2, 0);
+  VW::cats_tree::predict_test_helper({1}, 2, 2, 0);
+  // 4 node tree
+  VW::cats_tree::predict_test_helper({-1, 1}, 2, 4, 0);
+  VW::cats_tree::predict_test_helper({1, 1}, 4, 4, 0);
+  // 4 node tree with bandwidth 1
+  VW::cats_tree::predict_test_helper({-1}, 2, 4, 1);
+  VW::cats_tree::predict_test_helper({1}, 3, 4, 1);
+  // 8 node tree with bandwidth 1
+  VW::cats_tree::predict_test_helper({-1, -1}, 2, 8, 1);
+  VW::cats_tree::predict_test_helper({-1, 1, -1}, 3, 8, 1);
+  // 8 node tree with bandwidth 2
+  VW::cats_tree::predict_test_helper({-1, -1}, 3, 8, 2);
+  VW::cats_tree::predict_test_helper({1, 1}, 6, 8, 2);
+
+}
+
+BOOST_AUTO_TEST_CASE(build_min_depth_tree_cont_5)
+{
+  VW::cats_tree::min_depth_binary_tree tree;
+  tree.build_tree(4, 1);
+  std::vector<VW::cats_tree::tree_node> expected = {
+      {0, 1, 2, 0, 0, false, false, false},
+      {1, 3, 4, 0, 1, false, true, false},
+      {2, 5, 6, 0, 1, true, false, false},
+      {3, 0, 0, 1, 2, false, false, true},
+      {4, 0, 0, 1, 2, false, false, true},
+      {5, 0, 0, 2, 2, false, false, true},
+      {6, 0, 0, 2, 2, false, false, true},
+  };
+  BOOST_CHECK_EQUAL_COLLECTIONS(tree.nodes.begin(), tree.nodes.end(), expected.begin(), expected.end());
+}
+
+
+BOOST_AUTO_TEST_CASE(build_min_depth_tree_cont_1)
+{
+  VW::cats_tree::min_depth_binary_tree tree;
+  tree.build_tree(1, 0);
+  std::vector<VW::cats_tree::tree_node> expected = {
+      {0, 0, 0, 0, 0, false, false, true}
+  };
+  BOOST_CHECK_EQUAL_COLLECTIONS(tree.nodes.begin(), tree.nodes.end(), expected.begin(), expected.end());
+}
+
+BOOST_AUTO_TEST_CASE(build_min_depth_tree_cont_too_big, *boost::unit_test::disabled())
```

Comment:
Why is this test turned off?

Reply:
It is causing Valgrind to fail.  Valgrind cannot handle memory allocation failure test.

Reply:
Interesting, do you mean Valgrind interprets this as a failure but in reality we catch it?

Reply:
The test is expecting an exception to be thrown during allocation.  So the test catches the exception.  The issue is that there is a new failure which causes Valgrind to exit.

Reply:
Short answer is yes.